src/lib/components/Menu/Menu.js

```diff
@@ -122,6 +136,105 @@ class Menu extends Component {
     if (this.props.showMap && this.props.showMap !== prevProps.showMap && this.props.menuScroll) {
       this.menuWrapper.current.scrollTop = this.props.menuScroll.scrollTop;
     }
+
+    const { sharedLayers } = this.state;
+    if (sharedLayers.filter(l => !l.isCatOpen).length) {
+      /** If there are any shared layers whose category we haven't open,
+       * open them
+       */
+      this.openCategoryForSharedLayers(sharedLayers);
+    }
+  }
+
+  /**
+   * Open category for which each of the shared layers falls under
+   */
+  openCategoryForSharedLayers(layersToOpen) {
+    const { categories } = this.props;
+    // const { sharedLayers } = this.state;
+
+    if (layersToOpen && categories) {
+      layersToOpen
+        .filter(l => !l.isCatOpen)
+        .forEach(sharedLayer => {
+          let i = 0;
+          /**
+           * A layer belongs to only one category. So if we found its
+           * category, use this flag to break from the loop
+           */
+          let catFound = false;
+
+          while (!catFound && i < categories.length) {
+            const category = categories[i];
+            let j = 0;
+
+            while (!catFound && j < category.layers.length) {
+              const layer = category.layers[j];
+
+              if (!layer.id) {
+                /**
+                 * This is a group. So continue checking down the hierarchy
+                 * for visible layers
+                 */
+                const groupNames = Object.keys(layer);
+                let k = 0;
+
+                while (!catFound && k < groupNames.length) {
```

Comment:
@ciremusyoka this looks like quite a few nested loops...

I'm wondering if this could instead be a recursive function which will loop exactly as many times as necessary depending on how the categories/groups/layers are structured. The reason is because the menu can have any number of nested groups within a group, we typically limit it to 3 nests deep but it could easily be more just by restructuring the layers in the site-config.

Reply:
@cKellyDesign also this is from [map state to url](https://github.com/onaio/gisida-react/pull/195) too.

Reply:
@cKellyDesign method `getMenuGroupMapLayers` is recursive. It recursively checks the groups regardless of the nesting. So for each category, we check if it has one level (by checking `layer.id`) , if not then we proceed to check the nested groups